File: packages/examples/todo/backends/typescript/src/commands/create-batch.ts
/**
 * @fileoverview todo.createBatch command
 *
 * Demonstrates AFD batch operation pattern with partial failure handling.
 * Creates multiple todos at once, reporting success/failure for each item.
 */

import type { CommandError } from '@lushly-dev/afd-core';
import { defineCommand, success } from '@lushly-dev/afd-server';
import { z } from 'zod';
import { store } from '../store/index.js';
import type { Todo } from '../types.js';

/**
 * Schema for a single todo item in a batch
 */
const todoItemSchema = z.object({
	title: z.string().min(1, 'Title is required').max(200, 'Title too long'),
	description: z.string().max(1000).optional(),
	priority: z.enum(['low', 'medium', 'high']).default('medium'),
});

const inputSchema = z.object({
	todos: z
		.array(todoItemSchema)
		.min(1, 'At least one todo is required')
		.max(100, 'Maximum 100 todos per batch'),
});

/**
 * Result for a failed batch item
 */
export interface FailedItem {
	/** Position in input array (0-indexed) */
	index: number;
	/** The original input that failed */
	input: unknown;
	/** Why it failed */
	error: CommandError;
}

/**
 * Batch operation result
 */
export interface BatchCreateResult {
	/** Successfully created todos */
	succeeded: Todo[];
	/** Items that failed */
	failed: FailedItem[];
	/** Summary statistics */
	summary: {
		total: number;
		successCount: number;
		failureCount: number;
	};
}

export const createBatch = defineCommand<typeof inputSchema, BatchCreateResult>({
	name: 'todo-createBatch',
	description: 'Create multiple todos at once with partial failure support',
	category: 'todo',
	tags: ['todo', 'create', 'write', 'batch'],
	mutation: true,
	version: '1.0.0',
	input: inputSchema,
	errors: ['VALIDATION_ERROR', 'PARTIAL_FAILURE'],

	async handler(input) {
		const succeeded: Todo[] = [];
		const failed: FailedItem[] = [];

		// Process each todo item
		for (let i = 0; i < input.todos.length; i++) {
			const item = input.todos[i]!;

			try {
				// Validate individual item (already validated by Zod, but demonstrates pattern)
				if (!item.title || item.title.trim().length === 0) {
					failed.push({
						index: i,
						input: item,
						error: {
							code: 'VALIDATION_ERROR',
							message: 'Title is required',
							suggestion: 'Provide a non-empty title for this todo',
						},
					});
					continue;
				}

				// Create the todo
				const todo = store.create({
					title: item.title.trim(),
					description: item.description?.trim(),
					priority: item.priority,
				});

				succeeded.push(todo);
			} catch (err) {
				failed.push({
					index: i,
					input: item,
					error: {
						code: 'CREATE_ERROR',
						message: err instanceof Error ? err.message : 'Failed to create todo',
						suggestion: 'Check the input and try again',
					},
				});
			}
		}

		const summary = {
			total: input.todos.length,
			successCount: succeeded.length,
			failureCount: failed.length,
		};

		// Build reasoning based on results
		let reasoning: string;
		if (failed.length === 0) {
			reasoning = `Successfully created all ${succeeded.length} todos`;
		} else if (succeeded.length === 0) {
			reasoning = `Failed to create any todos. All ${failed.length} items had errors.`;
		} else {
			reasoning = `Created ${succeeded.length} of ${summary.total} todos. ${failed.length} failed validation.`;
		}

		// Calculate confidence based on success rate
		const confidence = summary.total > 0 ? summary.successCount / summary.total : 0;

		// Add warning if partial failure
		const warnings =
			failed.length > 0 && succeeded.length > 0
				? [
						{
							code: 'PARTIAL_SUCCESS',
							message: `${failed.length} of ${summary.total} items failed`,
							severity: 'warning' as const,
						},
					]
				: undefined;

		return success(
			{ succeeded, failed, summary },
			{
				reasoning,
				confidence,
				warnings,
			}
		);
	},
});

File: packages/examples/todo/backends/rust/src/commands/create_batch.rs
use super::{batch_confidence, ok, schema, warning, Command, FailedItem, Summary};
use crate::store::{NewTodo, TodoStore};
use crate::types::{Priority, Todo};
use afd::{CommandError, CommandResult, WarningSeverity};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;

pub struct CreateBatch(pub Arc<TodoStore>);

#[derive(Deserialize, Serialize)]
pub struct Item {
    title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    priority: Priority,
}

#[derive(Deserialize)]
pub struct Input {
    todos: Vec<Item>,
}

#[derive(Serialize)]
struct Output {
    succeeded: Vec<Todo>,
    failed: Vec<FailedItem>,
    summary: Summary,
}

impl Command for CreateBatch {
    type Input = Input;
    const NAME: &'static str = "todo-create-batch";
    const DESCRIPTION: &'static str = "Create multiple todos at once with partial failure support";
    const MUTATION: bool = true;

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": {
                "todos": {
                    "type": "array",
                    "minItems": 1,
                    "maxItems": 100,
                    "description": "Todos to create (1 to 100)",
                    "items": {
                        "type": "object",
                        "properties": {
                            "title": schema::title(),
                            "description": schema::description(),
                            "priority": schema::priority_with_default("Priority (default: medium)")
                        },
                        "required": ["title"]
                    }
                }
            },
            "required": ["todos"]
        })
    }

    fn run(&self, input: Input) -> CommandResult<Value> {
        let total = input.todos.len();
        let mut succeeded = Vec::new();
        let mut failed = Vec::new();

        for (index, item) in input.todos.into_iter().enumerate() {
            // The schema allows "   "; a title must still have visible text.
            let title = item.title.trim().to_string();
            if title.is_empty() {
                failed.push(FailedItem {
                    index,
                    id: None,
                    input: serde_json::to_value(&item).ok(),
                    error: CommandError::new("VALIDATION_ERROR", "Title is required")
                        .with_suggestion("Provide a non-empty title for this todo")
                        .with_retryable(false),
                });
                continue;
            }
            succeeded.push(self.0.create(NewTodo {
                title,
                description: item.description.map(|d| d.trim().to_string()),
                priority: item.priority,
            }));
        }

        let summary = Summary {
            total,
            success_count: succeeded.len(),
            failure_count: failed.len(),
        };
        let reasoning = if failed.is_empty() {
            format!("Successfully created all {} todos", succeeded.len())
        } else if succeeded.is_empty() {
            format!(
                "Failed to create any todos. All {} items had errors.",
                failed.len()
            )
        } else {
            format!(
                "Created {} of {total} todos. {} failed validation.",
                succeeded.len(),
                failed.len()
            )
        };
        let warnings = if !failed.is_empty() && !succeeded.is_empty() {
            vec![warning(
                "PARTIAL_SUCCESS",
                format!("{} of {total} items failed", failed.len()),
                WarningSeverity::Warning,
            )]
        } else {
            Vec::new()
        };
        let confidence = batch_confidence(summary.success_count, total);
        ok(
            &Output {
                succeeded,
                failed,
                summary,
            },
            reasoning,
            confidence,
            warnings,
        )
    }
}
